File: java/lib/src/main/java/com/svix/Application.java
package com.xwebhook;

import com.xwebhook.exceptions.ApiException;
import com.xwebhook.internal.api.ApplicationApi;
import com.xwebhook.models.ApplicationIn;
import com.xwebhook.models.ApplicationOut;
import com.xwebhook.models.ApplicationPatch;
import com.xwebhook.models.ListResponseApplicationOut;

public final class Application {
	private final ApplicationApi api;

	Application() {
		api = new ApplicationApi();
	}

	public ListResponseApplicationOut list(final ApplicationListOptions options) throws ApiException {
		try {
			return api.v1ApplicationList(options.getLimit(), options.getIterator(), options.getOrder());
		} catch (com.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public ApplicationOut create(final ApplicationIn applicationIn) throws ApiException {
		return this.create(applicationIn, new PostOptions());
	}

	public ApplicationOut create(final ApplicationIn applicationIn, final PostOptions options) throws ApiException {
		try {
			return api.v1ApplicationCreate(applicationIn, null, options.getIdempotencyKey());
		} catch (com.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public ApplicationOut getOrCreate(final ApplicationIn applicationIn) throws ApiException {
		return this.getOrCreate(applicationIn, new PostOptions());
	}

	public ApplicationOut getOrCreate(final ApplicationIn applicationIn, final PostOptions options) throws ApiException {
		try {
			return api.v1ApplicationCreate(applicationIn, true, options.getIdempotencyKey());
		} catch (com.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public ApplicationOut get(final String appId) throws ApiException {
		try {
			return api.v1ApplicationGet(appId);
		} catch (com.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public ApplicationOut update(final String appId, final ApplicationIn applicationIn) throws ApiException {
		try {
			return api.v1ApplicationUpdate(appId, applicationIn);
		} catch (com.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public ApplicationOut patch(final String appId, final ApplicationPatch applicationPatch) throws ApiException {
		try {
			return api.v1ApplicationPatch(appId, applicationPatch);
		} catch (com.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void delete(final String appId) throws ApiException {
		try {
			api.v1ApplicationDelete(appId);
		} catch (com.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}
}

File: kotlin/lib/src/main/kotlin/Application.kt
package com.xwebhook.kotlin

import com.xwebhook.kotlin.exceptions.ApiException
import com.xwebhook.kotlin.internal.apis.ApplicationApi
import com.xwebhook.kotlin.models.ApplicationIn
import com.xwebhook.kotlin.models.ApplicationOut
import com.xwebhook.kotlin.models.ApplicationPatch
import com.xwebhook.kotlin.models.ListResponseApplicationOut

class Application internal constructor(token: String, options: XwebhookOptions) {
    private val api = ApplicationApi(options.serverUrl)

    init {
        api.accessToken = token
        api.userAgent = options.getUA()
        options.initialRetryDelayMillis?.let { api.initialRetryDelayMillis = it }
        options.numRetries?.let { api.numRetries = it }
    }

    suspend fun list(options: ApplicationListOptions = ApplicationListOptions()): ListResponseApplicationOut {
        try {
            return api.v1ApplicationList(options.limit, options.iterator, options.order)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun create(
        applicationIn: ApplicationIn,
        options: PostOptions = PostOptions(),
    ): ApplicationOut {
        try {
            return api.v1ApplicationCreate(applicationIn, null, options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun getOrCreate(
        applicationIn: ApplicationIn,
        options: PostOptions = PostOptions(),
    ): ApplicationOut {
        try {
            return api.v1ApplicationCreate(applicationIn, true, options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun get(appId: String): ApplicationOut {
        try {
            return api.v1ApplicationGet(appId)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun update(
        appId: String,
        applicationIn: ApplicationIn,
    ): ApplicationOut {
        try {
            return api.v1ApplicationUpdate(appId, applicationIn)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun patch(
        appId: String,
        applicationPatch: ApplicationPatch,
    ): ApplicationOut {
        try {
            return api.v1ApplicationPatch(appId, applicationPatch)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun delete(appId: String) {
        try {
            api.v1ApplicationDelete(appId)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }
}
